### PR TITLE
Unify logsdb index settings providers

### PR DESCRIPTION
**Pull request by kkrik-es**
Follow-up on #115397, where it was suggested to merge the two providers, to avoid inter-dependencies for provided settings and simplify the logic.

### CODE REVIEW

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 1]
```diff
@@ -7,53 +7,114 @@
...
+        createdIndexVersion.set(indexVersion);
+    }
+
+    private boolean supportSyntheticSourceDemotion() {
+        return mapperServiceFactory.get() != null;
+    }
+
+    @Override
+    public boolean overrulesTemplateAndRequestSettings() {
+        // Indicates that the provider value takes precedence over any user setting.
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:67]: This means that both `index.mode` and `index.mapping.source.mode` will be overruling if returned by this settings provider.

**Review by martijnvg — COMMENTED** [review]
Looks good! 

But I think there is an issue if the new index settings provider returns `index.mode` setting and user template uses index.mode setting as well. In that case we would always overrule index.mode=logsdb, while the user may have opted out of logsdb?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 2]
```diff
@@ -7,53 +7,114 @@
...
+        Settings.Builder settingsBuilder = null;
         if (isLogsdbEnabled
             && dataStreamName != null
-            && resolveIndexMode(settings.get(IndexSettings.MODE.getKey())) == null
+            && resolveIndexMode(indexTemplateAndCreateRequestSettings.get(IndexSettings.MODE.getKey())) == null
             && matchesLogsPattern(dataStreamName)) {
-            return Settings.builder().put("index.mode", IndexMode.LOGSDB.getName()).build();
+            settingsBuilder = Settings.builder().put(IndexSettings.MODE.getKey(), IndexMode.LOGSDB.getName());
+            if (supportSyntheticSourceDemotion()) {
+                indexTemplateAndCreateRequestSettings = Settings.builder()
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:88]: This is copied from:

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/SyntheticSourceIndexSettingsProvider.java#L75

I suggest we keep this as pure refactoring, we can improve the logic separately.

Btw, don't we check for index.mode in settings in line 84 above?

**Comment by kkrik-es** [comment]
> But I think there is an issue if the new index settings provider returns `index.mode` setting and user template uses index.mode setting as well. In that case we would always overrule index.mode=logsdb, while the user may have opted out of logsdb?

I think the two settings are injected independently, there should be no change in the logic wrt when each gets applied. More so, `index.mode` is only injected if it's not present in the incoming settings, so the fact that this is an overruling provider shouldn't change much.

**Comment by martijnvg** [comment]
> More so, index.mode is only injected if it's not present in the incoming settings, so the fact that this is an overruling provider shouldn't change much.

Ah right, I missed that this: `&& resolveIndexMode(settings.get(IndexSettings.MODE.getKey())) == null` is checked before index.mode setting is injected.

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by kkrik-es** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/118475)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**